src/python/pants/backend/java/dependency_inference/rules.py

```diff
@@ -36,30 +40,61 @@ class InferJavaSourceDependencies(InferDependenciesRequest):
     infer_from = JavaSourceField
 
 
-@rule(desc="Inferring Java dependencies by analyzing imports")
-async def infer_java_dependencies_via_imports(
+@dataclass(frozen=True)
+class JavaInferredDependencies:
+    dependencies: FrozenOrderedSet[Address]
+    exports: FrozenOrderedSet[Address]
+
+
+@dataclass(frozen=True)
+class JavaInferredDependenciesAndExportsRequest:
+    address: Address
+
+
+@rule(desc="Inferring Java dependencies by source analysis")
+async def infer_java_dependencies_via_source_analysis(
     request: InferJavaSourceDependencies,
+) -> InferredDependencies:
+
+    jids = await Get(
+        JavaInferredDependencies,
+        JavaInferredDependenciesAndExportsRequest(request.sources_field.address),
+    )
+    return InferredDependencies(dependencies=jids.dependencies)
+
+
+@rule(desc="Inferring Java dependencies and exports by source analysis")
+async def infer_java_dependencies_and_exports_via_source_analysis(
+    request: JavaInferredDependenciesAndExportsRequest,
     java_infer_subsystem: JavaInferSubsystem,
     first_party_dep_map: FirstPartySymbolMapping,
     third_party_artifact_mapping: ThirdPartyPackageToArtifactMapping,
-) -> InferredDependencies:
+) -> JavaInferredDependencies:
     if (
         not java_infer_subsystem.imports
         and not java_infer_subsystem.consumed_types
         and not java_infer_subsystem.third_party_imports
     ):
-        return InferredDependencies([])
+        return JavaInferredDependencies(FrozenOrderedSet([]), FrozenOrderedSet([]))
+
+    address = request.address
+    if not address.is_file_target:
+        raise Exception(
+            "Java source analysis requires a snapshot with a "
+            f"single Java source file, but was provided with address `{address}`."
+        )
+    a = await Get(Digest, PathGlobs([address.filename]))
```

Comment:
Why not use `SourceFilesRequest` here?

Reply:
Over in `javac.py`, we need to request dependencies and exports, but only have `Address` available to us. `SourceFiles` ended up being difficult to use. I'm happy to take help direction if there's a better way to do this.

Reply:
Mm, yea: I missed this bit.

This will block codegen, but I'd be ok with opening a followup ticket for this, and/or commenting on the first relevant codegen ticket (probably #13469).

We will likely also need explicit support for `exports` by then, which will likely allow this method to have a cleaner implementation. Should open a new ticket for explicit exports, or keep https://github.com/pantsbuild/pants/issues/13553 open, since it has some discussion of how to do this. cc @Eric-Arellano

Reply:
I had a 5-min attempt at a refactor to achieve using `SourceFiles` yesterday, and got stuck, so I'll merge without, and add a new TODO ticket